Add information on how user binding for GitLab works (#573)
Enalean/tuleap-documentation-en: languages/en/user-guide/code-versioning/gitlab.rst

.. _version-control-with-gitlab:

GitLab plugin
=============

.. note::

  This plugin is under construction, the documentation and the plugin can change.

Overview
--------

The purpose of this integration is to link Tuleap projects and GitLab repositories.

If you are using GitLab and want to keep a trace of your commits in Tuleap, this plugin will
allow you to reference Tuleap artifacts in your commit messages and conversely.

References
----------

> A GitLab repository can be referenced in several Tuleap artifacts, in different projects.

Reviewer: We cannot reference a repository.
```suggestion
A GitLab commit can be referenced in several Tuleap artifacts, in different projects.
```

A Tuleap artifact can reference several commits, in different GitLab repositories.

Reference a Tuleap artifact
'''''''''''''''''''''''''''

To be able to create GitLab cross-references, you need to:

* Register your GitLab repository in the Git service of your Tuleap project
* Reference Tuleap artifacts in GitLab commit messages

To link your commit to the Tuleap artifact of your choice, you must add the keyword ``TULEAP-<artifact_id>`` (case-sensitive)
to your commit message. 

> You can reference as many artifacts as you want. However, you must ensure that the projects your artifacts

Reviewer: Are you sure that GitLab repositories and artifacts must be in the same project?

Reviewer 2: you are right, I lied :stuck_out_tongue:

> belong have your GitLab Repository registered.

Reviewer: I'm not sure about this one, but shouldn't we say "belong to"?
```suggestion
belong to have your GitLab repository registered.
```

When the committer email is matching a Tuleap account, then its avatar and username will be displayed in the reference.
Otherwise, the committer name is displayed as received from the Gitlab API.

Reference a GitLab commit in Tuleap
'''''''''''''''''''''''''''''''''''

Please refer to :ref:`reference-overview` for more details on references.

You can reference a commit of one of the GitLab repositories registered in your Tuleap project.

> To do so, you have to use the keyword ``gitlab_commit`` followed by a ``#`` and the commit URL:

Reviewer: It is not really the URL. You should use `sha1` or `id` instead of URL.
```suggestion
To do so, you have to use the keyword ``gitlab_commit`` followed by a ``#`` and the commit sha1:
```

.. figure:: ../../images/screenshots/gitlab/gitlab-api-scope.png
   :align: center
   :alt: GitLab API scope
   :name: GitLab API scope

The name of the token is not important, but you need at least to check ``api`` in the scopes list.

Once your GitLab API token is created, copy it and save it for later.

GitLab repository registration
''''''''''''''''''''''''''''''

Go to the Git service of your Tuleap project, click on [New repository], then click on [Add GitLab repository].

.. figure:: ../../images/screenshots/gitlab/button-gitlab-integration.png
   :align: center
   :alt: Button integrate GitLab
   :name: Button integrate GitLab

In the modal, provide the URL of your GitLab instance and your API token.

.. figure:: ../../images/screenshots/gitlab/modal-server-instance.png
   :align: center
   :alt: Modal to enter server instance and API token
   :name: Modal to enter server instance and API token

The list of the repositories you maintain is displayed. Select the repository to link.

.. figure:: ../../images/screenshots/gitlab/modal-choose-repository.png
   :align: center
   :alt: Modal to choose GitLab repository
   :name: Modal to choose GitLab repository

> Once the GitLab repository is registered, it is displayed in the repositories list and is visually identifiable thanks to a GitLab symbol.

Reviewer: Suggestion: GitLab icon?
```suggestion
Once the GitLab repository is registered, it is displayed in the repositories list and is visually identifiable thanks to the GitLab icon.
```

.. figure:: ../../images/screenshots/gitlab/tile-gitlab.png
   :align: center
   :alt: GitLab repository tile
   :name: GitLab repository tile

From now on, each time you reference an artifact in a commit, a cross-reference will be created in the target artifact.

.. note::
  During the registration, a "post push" webhook is created in the GitLab repository.
  If the parameters of this webhook change (URL, events, or anything else), we cannot ensure that cross-references will

>   be continue to be created.

Reviewer: ```suggestion
  continue to be created.
```

Unlinking repositories 
'''''''''''''''''''''''

> As a Git administrator, go to the Git service of your project and find the repository to be unregistered in the list. Then click on the trash icon in the top right corner of the tile. 

Reviewer: "to be unregistered": The title of the section is "Unlinking repositories". I suggest to use the same term.

.. figure:: ../../images/screenshots/gitlab/tile-gitlab.png
   :align: center
   :alt: unlink GitLab repository tile
   :name: unlink GitLab repository tile

Confirm the action.

.. figure:: ../../images/screenshots/gitlab/modal-confirm-unlink.png
   :align: center
   :alt: Modal to confirm unlink
   :name: Modal to confirm unlink

> From now on, existing references won't work anymore and any new commit in this repository referencing a Tuleap artifact in this project will create cross-references.

Reviewer: ```suggestion
From now on, existing references won't work anymore and any new commit in this repository referencing a Tuleap artifact in this project will not create cross-references.
```

Cross Reference is created only if there is an integration between the project of the referenced artifact and
GitLab repository of the commit.

> If the commit author's email is bind to an existing Tuleap user, then this user is displayed in the cross reference.

Reviewer: Shouldn't it be "is bound"?

```suggestion
If the commit author's email is bound to an existing Tuleap user, then this user is displayed in the cross reference.
```
